File: sdk/php/src/Connection/CliDownloader.php
<?php

namespace Dagger\Connection;

use Dagger\Exception\CliReleaseUnavailable;
use Psr\Log\LoggerAwareInterface;
use Psr\Log\LoggerInterface;
use Psr\Log\NullLogger;
use RuntimeException;

/**
 * @deprecated
 * dagger modules will always have the environment variables set
 * so we don't need to download a CLI Client
 */
class CliDownloader implements LoggerAwareInterface
{
    public const DAGGER_CLI_BIN_PREFIX = 'dagger-';

    private LoggerInterface $logger;

    public function __construct()
    {
        $this->logger = new NullLogger();
    }

    public function setLogger(LoggerInterface $logger): void
    {
        $this->logger = $logger;
    }

    public function download(?string $version = null): string
    {
        if (null === $version) {
            $version = Provisioning::getCliVersion();
        }

        $cacheDir = $this->getCacheDir();
        if (!file_exists($cacheDir)) {
            mkdir($cacheDir);
        }

        $binName = self::DAGGER_CLI_BIN_PREFIX . $version;
        if ($this->isWindows()) {
            $binName .= '.exe';
        }

        $binPath = $cacheDir . DIRECTORY_SEPARATOR . $binName;

        if (file_exists($binPath)) {
            return $binPath;
        }

        $tmpFile = tempnam($cacheDir, 'tmp-');
        $archiveName = $this->getDefaultCliArchiveName($version);
        $expectedChecksum = $this->getExpectedChecksum($version, $archiveName);
        $actualChecksum = $this->extractCli($archiveName, $version, $tmpFile);

        if ($expectedChecksum === $actualChecksum) {
            rename($tmpFile, $binPath);
            chmod($binPath, 0700);
        } else {
            throw new RuntimeException("Invalid checksum : {$actualChecksum}, expected : {$expectedChecksum}");
        }

        return $binPath;
    }

    private function getCacheDir(): string
    {
        $xdgCacheHome = getenv('XDG_CACHE_HOME');
        $rootCacheDir = false !== $xdgCacheHome ? $xdgCacheHome : sys_get_temp_dir();

        return $rootCacheDir . DIRECTORY_SEPARATOR . 'dagger';
    }

    private function isWindows(): bool
    {
        return 'windows' === $this->getOs();
    }

    protected function getDefaultCliArchiveName(string $version): string
    {
        $ext = $this->isWindows() ? 'zip' : 'tar.gz';

        return "dagger_v{$version}_{$this->getOs()}_{$this->getArch()}.{$ext}";
    }

    private function getArch(): string
    {
        $uname = php_uname('m');

        if (str_contains($uname, 'x86_64') || str_contains($uname, 'amd64')) {
            return 'amd64';
        } elseif (str_contains($uname, 'x86')) {
            return 'x86';
        } elseif (str_contains($uname, 'arm')) {
            return 'armv7';
        } elseif (str_contains($uname, 'aarch64')) {
            return 'arm64';
        } else {
            return 'unknown';
        }
    }

    private function getOs(): string
    {
        $os = strtolower(PHP_OS);

        if (str_contains($os, 'win')) {
            return 'windows';
        } elseif (str_contains($os, 'linux')) {
            return 'linux';
        } elseif (str_contains($os, 'darwin')) {
            return 'darwin';
        } else {
            return 'unknown';
        }
    }

    private function getExpectedChecksum(string $daggerVersion, string $archiveName): string
    {
        $checksumMapUrl = $this->getChecksumsUrl($daggerVersion);
        [$checksumMapContent, $statusCode] = $this->fetchUrl($checksumMapUrl);
        if (null !== $statusCode && $statusCode >= 400) {
            $message = "Failed to download checksums from {$checksumMapUrl}: HTTP {$statusCode}";
            if ($this->isCliReleaseUnavailable($statusCode)) {
                throw new CliReleaseUnavailable($message);
            }

            throw new RuntimeException($message);
        }
        if (false === $checksumMapContent) {
            throw new RuntimeException("Failed to download checksums from {$checksumMapUrl}");
        }

        foreach (explode("\n", trim($checksumMapContent)) as $checksumLine) {
            $parts = preg_split('/\s+/', trim($checksumLine), 2);
            if (false === $parts || 2 !== count($parts)) {
                throw new RuntimeException("Malformed checksum data from {$checksumMapUrl}");
            }

            [$checksum, $filename] = $parts;
            if ($filename === $archiveName) {
                if (1 !== preg_match('/^[a-f0-9]{64}$/i', $checksum)) {
                    throw new RuntimeException("Malformed checksum data from {$checksumMapUrl}");
                }

                return $checksum;
            }
        }

        throw new RuntimeException("Could not find checksum for {$archiveName}");
    }

    private function extractCli(string $archiveName, string $daggerVersion, string $tmpBinFile): string
    {
        $tmpArchiveFile = $this->getCacheDir() . DIRECTORY_SEPARATOR . $archiveName;
        $archiveUrl = $this->getArchiveUrl($daggerVersion, $archiveName);
        $this->logger->info("Downloading dagger {$daggerVersion} from {$archiveUrl}");
        [$archiveContent, $statusCode] = $this->fetchUrl($archiveUrl);
        if (null !== $statusCode && $statusCode >= 400) {
            throw new RuntimeException("Failed to download Dagger CLI archive from {$archiveUrl}: HTTP {$statusCode}");
        }
        if (false === $archiveContent) {
            throw new RuntimeException("Failed to download Dagger CLI archive from {$archiveUrl}");
        }
        file_put_contents($tmpArchiveFile, $archiveContent);

        if ($this->isWindows()) {
            throw new RuntimeException('Not implemented');
        } else {
            file_put_contents($tmpBinFile, file_get_contents("phar://{$tmpArchiveFile}/dagger"));
        }

        $archiveHash = hash_file('sha256', $tmpArchiveFile);

        unlink($tmpArchiveFile);

        return $archiveHash;
    }

    private function getChecksumsUrl(string $daggerVersion): string
    {
        return "https://dl.dagger.io/dagger/releases/{$daggerVersion}/checksums.txt";
    }

    private function getArchiveUrl(string $daggerVersion, string $archiveName): string
    {
        return "https://dl.dagger.io/dagger/releases/{$daggerVersion}/{$archiveName}";
    }

    /** @return array{string|false, int|null} */
    protected function fetchUrl(string $url): array
    {
        $context = stream_context_create([
            'http' => [
                'ignore_errors' => true,
            ],
        ]);
        $http_response_header = [];
        $content = @file_get_contents($url, false, $context);
        $responseHeaders = $http_response_header;
        $statusCode = null;

        foreach ($responseHeaders as $header) {
            if (1 === preg_match('/^HTTP\/\S+\s+(\d{3})\b/i', $header, $matches)) {
                $statusCode = (int) $matches[1];
            }
        }

        return [$content, $statusCode];
    }

    private function isCliReleaseUnavailable(int $statusCode): bool
    {
        // dl.dagger.io returns 403 for missing S3 objects.
        return 403 === $statusCode || 404 === $statusCode;
    }
}

File: sdk/java/dagger-java-sdk/src/main/java/io/dagger/client/engineconn/CLIDownloader.java
package io.dagger.client.engineconn;

import java.io.*;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.security.DigestInputStream;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.HashMap;
import java.util.HexFormat;
import java.util.Map;
import org.apache.commons.compress.archivers.ArchiveEntry;
import org.apache.commons.compress.archivers.ArchiveInputStream;
import org.apache.commons.compress.archivers.tar.TarArchiveInputStream;
import org.apache.commons.compress.archivers.zip.ZipArchiveInputStream;
import org.apache.commons.compress.compressors.gzip.GzipCompressorInputStream;
import org.freedesktop.BaseDirectory;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

class CLIDownloader {

  static final Logger LOG = LoggerFactory.getLogger(CLIDownloader.class);

  private static final File CACHE_DIR =
      Paths.get(BaseDirectory.get(BaseDirectory.XDG_CACHE_HOME), "dagger").toFile();

  private static final String DAGGER_CLI_BIN_PREFIX = "dagger-";

  private final FileFetcher fetcher;

  public CLIDownloader(FileFetcher fetcher) {
    this.fetcher = fetcher;
  }

  public CLIDownloader() {
    this(url -> new URL(url).openStream());
  }

  public String downloadCLI(String version) throws IOException {
    CACHE_DIR.mkdirs();
    CACHE_DIR.setExecutable(true, true);
    CACHE_DIR.setReadable(true, true);
    CACHE_DIR.setWritable(true, true);

    String binName = DAGGER_CLI_BIN_PREFIX + version;
    if (isWindows()) {
      binName += ".exe";
    }
    Path binPath = Paths.get(CACHE_DIR.getPath(), binName);

    if (!binPath.toFile().exists()) {
      downloadCLI(version, binPath);
    }

    return binPath.toString();
  }

  public String downloadCLI() throws IOException {
    return downloadCLI(Provisioning.getCLIVersion());
  }

  private void downloadCLI(String version, Path binPath) throws IOException {
    String binName = binPath.getFileName().toString();
    Path tmpBin = Files.createTempFile(CACHE_DIR.toPath(), "tmp-" + binName, null);
    try {
      String archiveName = getDefaultCLIArchiveName(version);
      String expectedChecksum = expectedChecksum(version, archiveName);
      if (expectedChecksum == null) {
        throw new IOException("Could not find checksum for " + archiveName);
      }
      String actualChecksum = extractCLI(archiveName, version, tmpBin);
      if (!actualChecksum.equals(expectedChecksum)) {
        throw new IOException("Checksum validation failed");
      }
      tmpBin.toFile().setExecutable(true);
      Files.move(tmpBin, binPath);
    } finally {
      Files.deleteIfExists(tmpBin);
    }
  }

  private String expectedChecksum(String version, String archiveName) throws IOException {
    Map<String, String> checksums = fetchChecksumMap(version);
    return checksums.get(archiveName);
  }

  private String getDefaultCLIArchiveName(String version) {
    String ext = isWindows() ? "zip" : "tar.gz";
    return String.format("dagger_v%s_%s_%s.%s", version, getOS(), getArch(), ext);
  }

  private Map<String, String> fetchChecksumMap(String version) throws IOException {
    Map<String, String> checksums = new HashMap<>();
    String checksumMapURL =
        String.format("https://dl.dagger.io/dagger/releases/%s/checksums.txt", version);
    try (BufferedInputStream in = new BufferedInputStream(fetcher.fetch(checksumMapURL))) {
      ByteArrayOutputStream out = new ByteArrayOutputStream();
      byte[] dataBuffer = new byte[1024];
      int bytesRead;
      while ((bytesRead = in.read(dataBuffer, 0, 1024)) != -1) {
        out.write(dataBuffer, 0, bytesRead);
      }
      BufferedReader reader =
          new BufferedReader(new StringReader(out.toString(StandardCharsets.UTF_8)));
      String line;
      while ((line = reader.readLine()) != null) {
        String[] arr = line.split("\\s+");
        checksums.put(arr[1], arr[0]);
      }
      return checksums;
    }
  }

  private String extractCLI(String archiveName, String version, Path dest) throws IOException {
    String cliArchiveURL =
        String.format("https://dl.dagger.io/dagger/releases/%s/%s", version, archiveName);
    LOG.info("Downloading Dagger CLI from " + cliArchiveURL);
    MessageDigest sha256;
    try {
      sha256 = MessageDigest.getInstance("SHA-256");
    } catch (NoSuchAlgorithmException nsae) {
      throw new IOException("Could not instantiate SHA-256 digester", nsae);
    }
    LOG.info("Extracting archive...");
    try (InputStream in =
        new BufferedInputStream(new DigestInputStream(fetcher.fetch(cliArchiveURL), sha256))) {
      if (isWindows()) {
        extractZip(in, dest);
      } else {
        extractTarGZ(in, dest);
      }
      byte[] checksum = sha256.digest();
      return HexFormat.of().formatHex(checksum);
    }
  }

  private void extractZip(InputStream in, Path dest) throws IOException {
    try (ZipArchiveInputStream zipIn = new ZipArchiveInputStream(in)) {
      extractCLIBin(zipIn, "dagger.exe", dest);
    }
  }

  private static void extractCLIBin(ArchiveInputStream in, String binName, Path dest)
      throws IOException {
    boolean found = false;
    ArchiveEntry entry;
    while ((entry = in.getNextEntry()) != null) {
      if (entry.isDirectory() || !binName.equals(entry.getName())) {
        continue;
      }
      int count;
      byte[] data = new byte[4096];
      FileOutputStream fos = new FileOutputStream(dest.toFile());
      try (BufferedOutputStream out = new BufferedOutputStream(fos, 4096)) {
        while ((count = in.read(data, 0, 4096)) != -1) {
          out.write(data, 0, count);
        }
      }
      found = true;
      break;
    }
    if (!found) {
      throw new IOException("Could not find dagger binary in CLI archive");
    }
  }

  private void extractTarGZ(InputStream in, Path dest) throws IOException {
    boolean found = false;
    try (GzipCompressorInputStream gzipIn = new GzipCompressorInputStream(in);
        TarArchiveInputStream tarIn = new TarArchiveInputStream(gzipIn)) {
      extractCLIBin(tarIn, "dagger", dest);
    }
  }

  private static boolean isWindows() {
    return System.getProperty("os.name").toLowerCase().contains("win");
  }

  private static String getOS() {
    String os = System.getProperty("os.name").toLowerCase();
    if (os.contains("win")) {
      return "windows";
    } else if (os.contains("linux")) {
      return "linux";
    } else if (os.contains("darwin") || os.contains("mac")) {
      return "darwin";
    } else {
      return "unknown";
    }
  }

  private static String getArch() {
    String arch = System.getProperty("os.arch").toLowerCase();
    if (arch.contains("x86_64") || arch.contains("amd64")) {
      return "amd64";
    } else if (arch.contains("x86")) {
      return "x86";
    } else if (arch.contains("arm")) {
      return "armv7";
    } else if (arch.contains("aarch64")) {
      return "arm64";
    } else {
      return "unknown";
    }
  }
}
